### PR TITLE
Add extra SANs to microcluster certificate

### PR DESCRIPTION
**Pull request by bschimke95**
Microcluster validates that the certificate SANs do match with the name that was created when creating the join token. 
This breaks the CAPI workflow as in that, the noden/hostname is unknown at the time the token is created. 

We can work around this issue by providing extraSANs to the microcluster certificate on init which is done in this PR.

See https://docs.google.com/document/d/1ZcEDgLwy_BroyJr_oqTOyqBsK0mvBt7EB6UCrCxC1mg/edit#heading=h.kavgppauvkj5 for more context.

### CODE REVIEW

File: src/k8s/pkg/k8sd/app/hooks_remove.go [hunk 1]
```diff
@@ -66,9 +66,11 @@ func (a *App) onPreRemove(ctx context.Context, s state.State, force bool) (rerr
 				log.Error(err, "Failed to create Kubernetes client", err)
 			}
 
-			log.Info("Deleting node from Kubernetes cluster")
-			if err := c.DeleteNode(ctx, s.Name()); err != nil {
-				log.Error(err, "Failed to remove Kubernetes node")
+			if c != nil {
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/app/hooks_remove.go:69]: @mateoflorido drive-by fix of KU-1270 as I was running into this issue while debugging

File: tests/integration/tests/test_clustering.py [hunk 2]
```diff
@@ -88,9 +88,42 @@ def test_no_remove(instances: List[harness.Instance]):
...
+        text=True,
+    )
+    join_token = out.stdout.strip()
+
+    join_config = """
+extra-sans:
+- my-token
+"""
+    joining_cp.exec(
+        ["k8s", "join-cluster", join_token, "--name", "my-node", "--file", "-"],
```
Comment (berkayoz) [inline tests/integration/tests/test_clustering.py:116]: We should also test not supplying a `--name` flag at all which is what we do in capi. The command will use the `hostname` if not supplied.

File: src/k8s/cmd/k8s/k8s_get_join_token.go [hunk 3]
```diff
@@ -21,6 +21,10 @@ func newGetJoinTokenCmd(env cmdutil.ExecutionEnvironment) *cobra.Command {
 		PreRun: chainPreRunHooks(hookRequireRoot(env)),
 		Args:   cmdutil.MaximumNArgs(env, 1),
 		Run: func(cmd *cobra.Command, args []string) {
+			if !opts.worker && len(args) == 0 {
+				cmd.PrintErrln("Error: A node name is required for control-plane nodes.")
```
Comment (HomayoonAlimohammadi) [inline src/k8s/cmd/k8s/k8s_get_join_token.go:25]: Nice! I think this is very useful. otherwise the MC error will cause a lot of confusion.

**Review by berkayoz — COMMENTED** [review]
Great work, looks good to me. One small comment.

File: src/k8s/pkg/k8sd/app/hooks_preinit.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		return nil
+	}
+
+	controlPlaneJoinConfig, err := utils.MicroclusterControlPlaneJoinConfigFromMap(initConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get control plane join config, boostrap %v: %w", bootstrap, err)
+	}
+	extraSANs := controlPlaneJoinConfig.ExtraSANS
+
+	err = os.Remove(filepath.Join(s.FileSystem().StateDir, "server.crt"))
```
Comment (HomayoonAlimohammadi) [inline src/k8s/pkg/k8sd/app/hooks_preinit.go:27]: nit: maybe we can do
```go
if err := ...; err != nil {
...
}
```

File: src/k8s/pkg/k8sd/app/hooks_preinit.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to remove server.crt: %w", err)
+	}
+
+	err = os.Remove(filepath.Join(s.FileSystem().StateDir, "server.key"))
```
Comment (HomayoonAlimohammadi) [inline src/k8s/pkg/k8sd/app/hooks_preinit.go:32]: here as well

File: src/k8s/pkg/k8sd/app/hooks_preinit.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		shared.CertOptions{
+			AddHosts:                true,
+			CommonName:              s.Name(),
+			SubjectAlternativeNames: extraSANs,
+		})
+	if err != nil {
+		return err
+	}
+
+	err = a.client.UpdateCertificate(ctx, microclusterTypes.ServerCertificateName, microclusterTypes.KeyPair{
```
Comment (HomayoonAlimohammadi) [inline src/k8s/pkg/k8sd/app/hooks_preinit.go:50]: and here

**Review by HomayoonAlimohammadi — APPROVED** [review]
Thanks a lot @bschimke95! Some minor comments, otherwise, LGTM!

File: src/k8s/pkg/k8sd/app/hooks_preinit.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+		return nil
+	}
+
+	controlPlaneJoinConfig, err := utils.MicroclusterControlPlaneJoinConfigFromMap(initConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get control plane join config, boostrap %v: %w", bootstrap, err)
+	}
+	extraSANs := controlPlaneJoinConfig.ExtraSANS
+
+	if err = os.Remove(filepath.Join(s.FileSystem().StateDir, "server.crt")); err != nil {
```
Comment (berkayoz) [inline src/k8s/pkg/k8sd/app/hooks_preinit.go:27]:
```suggestion
	if err := os.Remove(filepath.Join(s.FileSystem().StateDir, "server.crt")); err != nil {
```

File: src/k8s/pkg/k8sd/app/hooks_preinit.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+		return fmt.Errorf("failed to remove server.crt: %w", err)
+	}
+
+	if err = os.Remove(filepath.Join(s.FileSystem().StateDir, "server.key")); err != nil {
```
Comment (berkayoz) [inline src/k8s/pkg/k8sd/app/hooks_preinit.go:31]:
```suggestion
	if err := os.Remove(filepath.Join(s.FileSystem().StateDir, "server.key")); err != nil {
```

File: src/k8s/pkg/k8sd/app/hooks_preinit.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+		shared.CertOptions{
+			AddHosts:                true,
+			CommonName:              s.Name(),
+			SubjectAlternativeNames: extraSANs,
+		})
+	if err != nil {
+		return err
+	}
+
+	if err = a.client.UpdateCertificate(ctx, microclusterTypes.ServerCertificateName, microclusterTypes.KeyPair{
```
Comment (berkayoz) [inline src/k8s/pkg/k8sd/app/hooks_preinit.go:48]:
```suggestion
	if err := a.client.UpdateCertificate(ctx, microclusterTypes.ServerCertificateName, microclusterTypes.KeyPair{
```

**Review by berkayoz — APPROVED** [review]
LGTM, thank you!
